File: server/nodes.js
import { MATERIAL_NODES } from '../shared/materials.js';

const NODES_BY_ID = new Map(MATERIAL_NODES.map(node => [node.id, node]));

/**
 * Server-authoritative gather-node state. Depletion timestamps are persisted
 * in storage.state.nodes as { [nodeId]: depletedAt }; a node is depleted
 * while `now < depletedAt + respawnMs`, and its depletion entry is reaped
 * (deleted) once the interval has elapsed. Because the timestamp is absolute,
 * the remaining respawn time survives server restarts untouched.
 */
export class NodesManager {
  constructor(storage) {
    this.storage = storage;
    if (!this.storage.state.nodes || typeof this.storage.state.nodes !== 'object') {
      this.storage.state.nodes = {};
    }
    this.depletions = this.storage.state.nodes;
  }

  persist() {
    this.storage.state.nodes = this.depletions;
    this.storage.save();
  }

  isDepleted(nodeId, now = Date.now()) {
    const def = NODES_BY_ID.get(nodeId);
    if (!def) return false;
    const depletedAt = this.depletions[nodeId];
    if (depletedAt === undefined || depletedAt === null) return false;
    return now < depletedAt + def.respawnMs;
  }

  /** District a node belongs to, or null for unknown ids. */
  nodeDistrict(nodeId) {
    const def = NODES_BY_ID.get(nodeId);
    return def ? def.district : null;
  }

  respawnAt(nodeId) {
    const def = NODES_BY_ID.get(nodeId);
    const depletedAt = this.depletions[nodeId];
    if (!def || depletedAt === undefined || depletedAt === null) return null;
    return depletedAt + def.respawnMs;
  }

  /**
   * Harvests a node: rejects unknown or still-depleted nodes without side
   * effects, otherwise marks it depleted at `now` and reports the material
   * yielded. The caller grants the material to the player inside the same
   * server-validated action.
   */
  harvest(nodeId, now = Date.now()) {
    const def = NODES_BY_ID.get(nodeId);
    if (!def) return { success: false, reason: 'unknown_node' };
    if (this.isDepleted(nodeId, now)) {
      return { success: false, reason: 'node_depleted', respawnAt: this.respawnAt(nodeId) };
    }
    this.depletions[nodeId] = now;
    this.persist();
    return { success: true, nodeId, material: def.material, district: def.district };
  }

  /**
   * Deletes (reaps) depletion entries whose respawn interval has elapsed.
   * Returns true when at least one entry was reaped.
   */
  reapExpired(now = Date.now()) {
    let reaped = false;
    for (const nodeId of Object.keys(this.depletions)) {
      const def = NODES_BY_ID.get(nodeId);
      const depletedAt = this.depletions[nodeId];
      if (!def || depletedAt === undefined || now >= depletedAt + def.respawnMs) {
        delete this.depletions[nodeId];
        reaped = true;
      }
    }
    if (reaped) this.persist();
    return reaped;
  }

  /**
   * Snapshot of every node in a district for wire transfer. Reaps expired
   * entries first so the snapshot reflects respawns. Returns null for
   * districts without nodes.
   */
  getStatesForDistrict(districtId, now = Date.now()) {
    const defs = MATERIAL_NODES.filter(node => node.district === districtId);
    if (defs.length === 0) return null;
    this.reapExpired(now);
    return defs.map(def => ({
      nodeId: def.id,
      material: def.material,
      available: !this.isDepleted(def.id, now),
      depletedAt: this.isDepleted(def.id, now) ? this.depletions[def.id] : null,
      respawnAt: this.respawnAt(def.id),
    }));
  }

  /**
     * Advances respawn timers and returns the ids of districts where a node
     * transitioned from depleted to available, so the caller can broadcast
     * NODE_STATE to the players standing there.
     */
  tick(now = Date.now()) {
    const changed = [];
    for (const districtId of new Set(MATERIAL_NODES.map(node => node.district))) {
      let districtChanged = false;
      for (const node of MATERIAL_NODES) {
        if (node.district !== districtId) continue;
        const depletedAt = this.depletions[node.id];
        if (depletedAt !== undefined && depletedAt !== null && now >= depletedAt + node.respawnMs) {
          delete this.depletions[node.id];
          districtChanged = true;
        }
      }
      if (districtChanged) changed.push(districtId);
    }
    if (changed.length > 0) this.persist();
    return changed;
  }
}

File: server_elixir/lib/afterlight/restoration/nodes.ex
defmodule Afterlight.Restoration.Nodes do
  @moduledoc """
  Gather-node depletion logic from `server/nodes.js`.
  """

  alias Afterlight.Restoration.Materials

  def fresh_state do
    %{"storage" => %{"state" => %{"nodes" => %{}}}, "depletions" => %{}}
  end

  def is_depleted?(state, node_id, now) do
    case Materials.node(node_id) do
      nil -> false
      def ->
        case state["depletions"][node_id] do
          nil -> false
          depleted_at -> now < depleted_at + def.respawn_ms
        end
    end
  end

  def respawn_at(state, node_id) do
    case Materials.node(node_id) do
      nil ->
        nil

      def ->
        case state["depletions"][node_id] do
          nil -> nil
          depleted_at -> depleted_at + def.respawn_ms
        end
    end
  end

  def harvest(state, node_id, now) do
    case Materials.node(node_id) do
      nil ->
        {state, %{"success" => false, "reason" => "unknown_node"}}

      def ->
        if is_depleted?(state, node_id, now) do
          {state,
           %{
             "success" => false,
             "reason" => "node_depleted",
             "respawnAt" => respawn_at(state, node_id)
           }}
        else
          state = put_depletions(state, Map.put(state["depletions"], node_id, now))

          {state,
           %{
             "success" => true,
             "nodeId" => def.id,
             "material" => def.material,
             "district" => def.district
           }}
        end
    end
  end

  def reap_expired(state, now) do
    depletions =
      Enum.reduce(state["depletions"], state["depletions"], fn {node_id, depleted_at}, acc ->
        def = Materials.node(node_id)

        if def == nil or now >= (depleted_at || 0) + def.respawn_ms do
          Map.delete(acc, node_id)
        else
          acc
        end
      end)

    put_depletions(state, depletions)
  end

  def states_for_district(state, district, now) do
    defs = Enum.filter(Materials.material_nodes(), &(&1.district == district))

    if defs == [] do
      nil
    else
      state = reap_expired(state, now)

      Enum.map(defs, fn def ->
        depleted = is_depleted?(state, def.id, now)

        %{
          "nodeId" => def.id,
          "material" => def.material,
          "available" => not depleted,
          "depletedAt" => if(depleted, do: state["depletions"][def.id], else: nil),
          "respawnAt" => respawn_at(state, def.id)
        }
      end)
    end
  end

  def tick(state, now) do
    districts =
      Materials.material_nodes()
      |> Enum.map(& &1.district)
      |> Enum.uniq()

    changed =
      Enum.reduce(districts, [], fn district_id, acc ->
        district_changed =
          Materials.material_nodes()
          |> Enum.filter(&(&1.district == district_id))
          |> Enum.any?(fn node ->
            depleted_at = state["depletions"][node.id]

            depleted_at != nil and now >= depleted_at + node.respawn_ms
          end)

        if district_changed do
          [district_id | acc]
        else
          acc
        end
      end)

    state =
      if changed != [] do
        depletions =
          Enum.reduce(state["depletions"], %{}, fn {node_id, depleted_at}, acc ->
            def = Materials.node(node_id)

            if def != nil and depleted_at != nil and now >= depleted_at + def.respawn_ms do
              acc
            else
              Map.put(acc, node_id, depleted_at)
            end
          end)

        put_depletions(state, depletions)
      else
        state
      end

    {state, changed}
  end

  defp put_depletions(state, depletions) do
    state
    |> put_in(["storage", "state", "nodes"], depletions)
    |> Map.put("depletions", depletions)
  end
end
